packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts

```diff
@@ -98,154 +98,123 @@ export class MenusContributionPointHandler {
-                        if (group) {
-                            action.group = group + (action.group ? '/' + action.group.split('@')[0] : '/_');
+                    if (menu.isSubmenu) {
+                        let [submenuGroup, submenuOrder = ''] = (menu.group || '_').split('@');
+                        // Generating group in format: `<submenu group>/<submenu name>`
+                        submenuGroup = `${submenuGroup}/${menu.label}`;
+                        if (submenusOrder) {
+                            // Adding previous submenus order to the start of current submenu order
+                            // in format: `<submenu order>/<submenu order>.../<menu order>`.
+                            submenuOrder = `${submenusOrder}/${submenuOrder}`;
                         }
-                        if (action.submenu) {
-                            const submenu: Submenu = allSubmenus!.find(s => s.id === action.submenu)!;
-                            registerActions(allMenus[action.submenu], action.group!.split('@')[0] + '/' + submenu.label);
+                        registerMenuActions(menu.children, submenuGroup, submenuOrder);
+                    } else {
+                        menu.submenusOrder = submenusOrder;
+                        toDispose.push(
+                            this.registerAction(plugin, rootMenu.id!, menu)
+                        );
+                    }
+                });
+            };
+
+            registerMenuActions(rootMenu.children, undefined, undefined);
+        });
+
+        return toDispose;
+    }
+
+    /**
+     * Transforms the structure of Menus & Submenus
+     * into something more tree-like.
+     */
+    protected getMenusTree(plugin: DeployedPlugin): MenuTree[] {
+        const allMenus = plugin.contributes && plugin.contributes.menus;
+        if (!allMenus) {
+            return [];
+        }
+        const allSubmenus = plugin.contributes && plugin.contributes.submenus;
+        const tree: MenuTree[] = [];
+
+        Object.keys(allMenus).forEach(location => {
+            // Don't build menus tree for a submenu decleration at root.
+            if (allSubmenus && allSubmenus.findIndex(submenu => submenu.id === location) > -1) {
+                return;
+            }
+
+            /**
+             * @param menus the menus to create a tree from.
+             * @param submenusIds contains all the previous submenus ids in the current tree.
+             * @returns {MenuTree[]} the trees for the given menus.
+             */
+            const getChildren = (menus: Menu[], submenusIds: Set<string>) => {
+                // Contains all the submenus ids of the current parent.
+                const parentSubmenusIds = new Set<string>();
+
+                return menus.reduce((children: MenuTree[], menuItem) => {
+                    if (menuItem.submenu) {
+                        if (parentSubmenusIds.has(menuItem.submenu)) {
```

Comment:
Hi @colin-grant-work I updated the code because `parentSubmenusIds ` declaration  was inside the `reduce`
so it didn't work as excepted, I found by chance that the menu was duplicated...